### DISCUSSION TITLE
데이터로거 첫 펌웨어 업로드 후 SRST 불가 에러메시지

### DISCUSSION THREAD

**Discussion by ghost** [luftaquila/monolith#15] [Q&A] [answered]
안녕하세요,
luftaquila님께서 올리신 데이터로거에 관심이 있어
몇 달 전 컨택을 하고 데이터로거를 제작한 국민대 kust팀입니다.
다름이 아니라, 데이터로거 하드웨어 제작 후 슬롯에 sd카드를 꽂은 채로 기본적인 기능(sd card, release debug)만 활성화한 채 펌웨어를 업로드하고 콘솔을 봤는데, 다음의 에러 메시지가 뜨는 것을 확인했습니다.

INFO: build finished. start flashing...

INFO: flashing TMA-1 STM32 binary...
Open On-Chip Debugger 0.12.0 (2023-07-12) [https://github.com/sysprogs/openocd]
Licensed under GNU GPL v2
libusb1 09e75e98b4d9ea7909e8837b7a3f00dda4589dc3
For bug reports, read
        http://openocd.org/doc/doxygen/bugs.html
Info : STLINK V2J37S7 (API v2) VID:PID 0483:3748
Info : Target voltage: 3.263298
Info : Unable to match requested speed 8000 kHz, using 4000 kHz
Info : Unable to match requested speed 8000 kHz, using 4000 kHz
Info : clock speed 4000 kHz
Error: BUG: can't assert SRST
Info : stlink_dap_op_connect(connect)
Error: BUG: can't assert SRST
Info : SWD DPIDR 0x2ba01477
Info : [STM32F407VETx.cpu] Cortex-M4 r0p1 processor detected
Info : [STM32F407VETx.cpu] target has 6 breakpoints, 4 watchpoints
Info : starting gdb server for STM32F407VETx.cpu on 3333
Info : Listening on port 3333 for gdb connections
Warn : [STM32F407VETx.cpu] target was in unknown state when halt was requested
[STM32F407VETx.cpu] halted due to debug-request, current mode: Thread
xPSR: 0x01000000 pc: 0x08008bfc msp: 0x20020000
configuring PLL
Info : Unable to match requested speed 8000 kHz, using 4000 kHz
Info : Unable to match requested speed 8000 kHz, using 4000 kHz
** Programming Started **
Info : device id = 0x10076413
Info : flash size = 512 KiB
** Programming Finished **
** Verify Started **
** Verified OK **
** Resetting Target **
shutdown command invoked
INFO: TMA-1 STM32 binary successfully flashed. please recyle the power.

INFO: TMA-1 STM32 flashing finished.

위의 에러 메시지가 뜨면서 콘솔 내용을 참고했을 때, stm 프로젝트 중 main문의 흐름상
  /* Configure the system clock */
  SystemClock_Config();

  /* USER CODE BEGIN SysInit */

위의 부분을 벗어나지 못한 것으로 보입니다.

이 중 Info : Unable to match requested speed 8000 kHz, using 4000 kHz
Info : Unable to match requested speed 8000 kHz, using 4000 kHz
Info : clock speed 4000 kHz 라는 부분이 있던데, ioc clock configuration도 원본 프로젝트를 참고해서 똑같이 설정하였습니다.
혹시 이런 문제를 겪은 적이 있으신지 궁금해서 남겨봅니다.
![image](https://github.com/luftaquila/monolith/assets/127092705/a17e837a-808a-4c3a-98b6-2a4b7d184b74)

**Comment by luftaquila** [comment]
안녕하세요, 답변이 늦어서 죄송합니다.

```
** Programming Started **
Info : device id = 0x10076413
Info : flash size = 512 KiB
** Programming Finished **
** Verify Started **
** Verified OK **
** Resetting Target **
```
올려주신 로그 내용으로 봐서는 정상적으로 업로드된 것 같습니다.

```
Info : Unable to match requested speed 8000 kHz, using 4000 kHz
Info : Unable to match requested speed 8000 kHz, using 4000 kHz
```

부분은 크게 개의치 않으셔도 됩니다. 업로드 과정에는 영향이 없습니다.

**Reply by luftaquila** [reply]
참고로 업로드가 완료되면 TMA-1은 다음 부팅 시부터 데이터로깅을 수행하지만, 동작 중에 출력을 따로 내지는 않습니다.

데이터로거가 정상 동작하는지 확인하는 방법은 다음과 같습니다.

* PCB의 HEARTBEAT LED가 점멸하는지 확인
* SD 카드를 TMA-1에 꽂고 전원을 켠 다음, SD카드를 뽑아 컴퓨터에서 로그 파일이 생성되었는지 확인
* 설정 도구에서 Debug mode를 켠 상태로 펌웨어를 업로드한 다음, PCB의 DEBUG UART 포트에 USB to UART 컨버터를 연결하고 시리얼 모니터 출력을 확인

**Reply by ghost** [reply, marked as answer]
답장이 늦었네요.
TMA-1.cfg 파일을 다음과 같이 수정했습니다. 한마디로 강제로 srst를 실행하게 하여 flash까지 실행했습니다.
![image](https://github.com/luftaquila/monolith/assets/127092705/d2beceb1-acc7-44be-8a4c-d29355078fb5)

sd 기능만 키고 업로드 해봤는데 콘솔은 다음과 같이 떴습니다. uart랑 led로 확인해보니까 정상 업로드 되네요.
![image](https://github.com/luftaquila/monolith/assets/127092705/0b3d6970-3bd4-43f4-8dcf-bb6cb6884b65)


감사합니다. 늘 도움 많이 받고 있습니다.
하시는 일 다 잘되시길!

**Reply by luftaquila** [reply]
해결되었다니 다행이네요. 사용하시면서 겪으시는 문제점이나 궁금한 점 많이 공유해 주세요. 감사합니다 😄
